### DISCUSSION TITLE
Adding and creating a sime shell for the os?

### DISCUSSION THREAD

**Discussion by 0xC0ba1t** [phil-opp/blog_os#1170] [Questions]
Hi guys,
I wanted to expand and add a shell with some simple games with the os I created following the series, I know they are planned in edition 3 but is there a way to use it with edition 2, because only booting and the freestanding binary posts are completed, thanks.

**Comment by phil-opp** [comment]
Hi, the basic steps could be:

- Follow the blog until at least the [Hardware Interrupts](https://os.phil-opp.com/hardware-interrupts/) post because you need keyboard input for creating a shell.
- In addition to printing characters to the screen, add them to some kind of string buffer (e.g. a `[char]` array, or an `alloc::string::String` if you completed the [Heap Allocation](https://os.phil-opp.com/heap-allocation/) post already).
- On a newline character, match the string buffer against some possible commands.
- If it matches a command, run it and print the results. Then wait for the next input.

For more advanced command-line input (e.g. arguments and `--flags`), I recommend to use one of the many existing command-line parsing libraries on crates.io. For example the [`clap`](https://docs.rs/clap/latest/clap/) crate should be `no_std` compatible with `default-features = false`.
